pyramid/tierFS.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	tempPath := tfs.workspaceFilePath(namespace, filename)
+	fh, err := os.Create(tempPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating file: %w", err)
+	}
+
+	return &File{
+		fh:       fh,
+		readOnly: false,
+		eviction: tfs.eviction,
+		store: func() error {
+			return tfs.store(namespace, tempPath, filename)
+		},
+	}, nil
+}
+
+// Load returns the a file descriptor to the local file.
+// If the file is missing from the local disk, it will try to fetch it from the block storage.
+func (tfs *TierFS) Open(namespace, filename string) (*File, error) {
+	fileRef := tfs.newLocalFileRef(namespace, filename)
+	fh, err := os.Open(fileRef.fullPath)
+	if err == nil {
+		return tfs.openFile(fileRef, fh)
+	}
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("open file: %w", err)
+	}
+
+	fh, err = tfs.readFromBlockStorage(fileRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return tfs.openFile(fileRef, fh)
+}
+
+// openFile converts an os.File to pyramid.File and updates the eviction control.
+func (tfs *TierFS) openFile(fileRef localFileRef, fh *os.File) (*File, error) {
+	stat, err := fh.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("file stat: %w", err)
+	}
+
+	// No need to check for the store output here,
+	// we already have the file.
+	tfs.eviction.store(fileRef.fsRelativePath, stat.Size())
+	return &File{
+		fh:       fh,
+		readOnly: true,
+		rPath:    fileRef.fsRelativePath,
+		eviction: tfs.eviction,
+	}, nil
+}
+
+// readFromBlockStorage reads the referenced file from the block storage
+// and places it in the local FS for further reading.
+// It returns a file handle to the local file.
+func (tfs *TierFS) readFromBlockStorage(fileRef localFileRef) (*os.File, error) {
+	reader, err := tfs.adaptor.Get(tfs.objPointer(fileRef.namespace, fileRef.filename), 0)
+	if err != nil {
+		return nil, fmt.Errorf("read from block storage: %w", err)
+	}
+	defer reader.Close()
+
+	writer, err := os.Create(fileRef.fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating file: %w", err)
+	}
+
+	if _, err := io.Copy(writer, reader); err != nil {
+		return nil, fmt.Errorf("copying date to file: %w", err)
+	}
+	if err := writer.Close(); err != nil {
+		return nil, fmt.Errorf("writer close: %w", err)
+	}
+
+	fh, err := os.Open(fileRef.fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %w", err)
+	}
+	return fh, nil
+}
+
+func (tfs *TierFS) validateArgs(namespace, filename string) error {
+	if err := tfs.validateNamespace(namespace); err != nil {
+		return err
+	}
+
+	return tfs.validateFilename(filename)
+}
+
+func (tfs *TierFS) validateFilename(filename string) error {
+	return tfs.containsSeparator(filename)
+}
+
+func (tfs *TierFS) validateNamespace(ns string) error {
+	return tfs.containsSeparator(ns)
+}
+
+var errSeparatorInFS = errors.New("path contains separator")
+
+func (tfs *TierFS) containsSeparator(str string) error {
+	if strings.ContainsRune(str, os.PathSeparator) {
```

Comment:
When called from `validateFilename` this doesn't work for Windows :vomiting_face: (I think), because S3 pathnames use `/` as a separator and Windows uses `\` :shrug:.